lib/rgeo/cartesian/planar_graph.rb

```diff
@@ -0,0 +1,400 @@
+# frozen_string_literal: true
+
+module RGeo
+  module Cartesian
+    module Graphs
+      # HalfEdge represents an edge as 2 directed edges.
+      # One half-edge will have it's origin at edge.s, the other
+      # at edge.e. Both half-edges will be linked as each other's twins.
+      #
+      # HalfEdges also contain references to the next and prev half-edges,
+      # where next's origin is this half-edges destination. Prev's destination
+      # is this half-edge's origin.
+      class HalfEdge
+        include Comparable
+
+        # Creates 2 half edges from an edge.
+        # They will be assigned as each other's twins.
+        # The Half Edges will be returned in the order of points
+        # of the edge (start, end).
+        #
+        # @param edge [RGeo::Cartesian::Segment]
+        #
+        # @return [Array]
+        def self.from_edge(edge)
+          # TODO: should we raise an error if the edge is degenerate?
+          e1 = new(edge.s)
+          e2 = new(edge.e)
+
+          e1.twin = e2
+          e2.twin = e1
+          [e1, e2]
+        end
+
+        def initialize(origin)
+          @origin = origin
+          @twin = nil
+          @next = nil
+          @prev = nil
+        end
+        attr_reader :origin
+        attr_accessor :twin, :next, :prev
+
+        # HalfEdges will be sorted around their shared vertex
+        # in a CW fashion. This means that face interiors will be
+        # a CCW.
+        def <=>(other)
+          angle <=> other.angle
+        end
+
+        # Will keep following next in a half-edge until it returns to itself
+        # or hits a half-edge without a next edge.
+        #
+        # @return [Array]
+        def each
+          hedges = []
+          yield(self) if block_given?
+          hedges << self
+
+          n = self.next
+          until n.eql?(self) || n.nil?
+            yield(n) if block_given?
+            hedges << n
+            n = n.next
+          end
+          hedges
+        end
```

Comment:
This might not be the best name for the method.

Also, it could be possible that this ends up in an infinite somehow. I could add a counter and if the counter exceeds the number of vertices or edges, then it should return what it is has or raise an error.

Reply:
Maybe I am overlooking something, but couldn't hedges be a set, and we would check for already existing hedge in this set each time ?

Also, why do you feel like each is not a good name ? 

One issue I could think of is that we're calling it on an edge, yet it returns the whole graph containing that edge. If that is correct, could it be `graph` or `related_hedges` ?

Reply:
> Maybe I am overlooking something, but couldn't hedges be a set, and we would check for already existing hedge in this set each time ?

Yes, it probably could be, but would it be weird to return a set? I guess it doesn't matter since a set is still enumerable.

> Also, why do you feel like each is not a good name ?

> One issue I could think of is that we're calling it on an edge, yet it returns the whole graph containing that edge. If that is correct, could it be graph or related_hedges ?

That's exactly why I thought `each` might not make sense. It's not like `HalfEdge` is a collection, so calling each on it and returning a bunch of hedges seemed weird since `each` is generally reserved for collections/enumerables. I was thinking `cycle`, but `related_hedges` might be clearer.

Reply:
I changed it to a Set which I think is better and renamed the method to `cycle`. `cycle` is already an `Enumerable` method and I think it describes the operation better.

Reply:
`linked_edges` maybe? As cycle is also standard lib it might still be a bit confusing.

Although, we could make this protected and be sure users would not call this method anyway